### PR TITLE
Support matmul semanthics for WMMA dot operation

### PR DESCRIPTION
**Pull request by joviliast**
-Add operand types checking
-Add result types conversion depending on the operands

Signed-off-by: joviliast <iveselov.nn@gmail.com>

### CODE REVIEW

File: python/triton/language/semantic.py [hunk 1]
```diff
@@ -1208,10 +1208,39 @@ def mfma_supported(M, N, K, allow_tf32, ret_scalar_ty, target) -> bool:
         return False
     return True
 
+def mfma_supported_granularity(m, n, k) -> bool:
```
Comment (alefimov-amd) [inline python/triton/language/semantic.py:1211]: maybe it should look like this?
```suggestion
def wmma_supported_granularity(m, n, k) -> bool:
```

File: python/triton/language/semantic.py [hunk 2]
```diff
@@ -1208,10 +1208,39 @@ def mfma_supported(M, N, K, allow_tf32, ret_scalar_ty, target) -> bool:
...
+    if lhs_dtype.scalar != rhs_dtype.scalar:
+        return False
+
+    if not lhs_dtype.scalar.is_uint8() and \
+       not lhs_dtype.scalar.is_bf16() and \
+       not lhs_dtype.scalar.is_fp16():
+       return False
+
+    # TODO: support some kind of padding
+    if M % 16 != 0 or N % 16 != 0 or K % 16 != 0:
```
Comment (alefimov-amd) [inline python/triton/language/semantic.py:1238]: I think you added "granularity" checked for this.

File: python/triton/language/semantic.py [hunk 3]
```diff
@@ -1306,6 +1323,33 @@ def assert_dtypes_valid(lhs_dtype, rhs_dtype, target):
...
+            _0 = builder.create_splat(builder.get_int32(0), [M, N])
+        elif lhs.type.scalar.is_fp16:
+            if ret_scalar_ty.is_fp16:
+                ret_dot_scalar_ty = tl.float16
+                _0 = builder.create_splat(builder.get_fp16(0), [M, N])
+            else:
+                ret_dot_scalar_ty = tl.float32
+                _0 = builder.create_splat(builder.get_fp32(0), [M, N])
+        elif lhs.type.scalar.is_bf16:
+            print(lhs.type.scalar)
```
Comment (chicheng) [inline python/triton/language/semantic.py:1341]: `print()`

**Comment by alefimov-amd** [comment]
I don't think this PR will be merged, because it will prevents backend integration in upstream.

**Comment by zhanglx13** [comment]
@alefimov-amd Agreed
It's better to push such checks in the backend. I remembered you are thinking of doing so for mfma instructions. Maybe you can work with @joviliast to figure out a better way of doing it.

**Comment by zhanglx13** [comment]
closing since wmma code are upstream'ed

**Pull request closed without merging**